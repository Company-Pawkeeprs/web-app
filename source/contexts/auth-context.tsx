'use client'

import { useRouter } from 'next/navigation';
import { createContext, useEffect } from "react";
import cookies from '~/constants/cookies';
import LOADING from '~/constants/loading';
import { decrypt, encrypt } from '~/helpers/encrypt-and-decrypt';
import {
    recoverUserByToken,
    signInUser
} from '~/store/auth/login/actions';
import {
    LoginState,
    onChangePassword,
    onChangeRememberMe,
    onChangeUsername,
    onSetRememberMe,
    onToggleVisiblePassword as onToggleVisiblePasswordAction,
} from '~/store/auth/login/slice';
import { useAppDispatch, useAppSelector } from '~/store/hooks';
import { getCookie, setCookie } from '~/utils/cookies-utils';

interface SignInData {
    username: string;
    password: string;
}

interface AuthContextType {
    isAuthenticated: boolean;
    user: any;
    password: string;
    username: string;
    visiblePassword: boolean;
    isLoading: LOADING;
    rememberMe: boolean;
    onToggleVisiblePassword: () => void;
    onToggleRememberMe: () => void;
    signIn: (data: SignInData) => Promise<void>;
}

export const AuthContext = createContext<AuthContextType>({} as AuthContextType);

interface AuthProviderProps {
    children: React.ReactNode;
}


export function AuthProvider({ children }: AuthProviderProps) {
    const dispatch = useAppDispatch()
    const {
        user,
        isAuthenticated,
        isLoading,
        password,
        rememberMe,
        username,
        visiblePassword
    } = useAppSelector(state => state.Login as LoginState)
    const router = useRouter()

    useEffect(() => {
        const token = getCookie(cookies.token.name)

        if (!token) {
            router.prefetch('/sign-in')
            return
        }

        dispatch(recoverUserByToken(token))

        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [])

    useEffect(() => {
        getRememberInfo()
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [])

    async function signIn({ username, password }: SignInData) {
        await setRememberInfo()
        dispatch(signInUser({ username, password }))
    }

    async function setRememberInfo() {
        const JSON_REMEMBER = JSON.stringify({ username, password: encrypt(password) })

        if (rememberMe) {
            setCookie(cookies.remember.name, JSON_REMEMBER, cookies.remember.expires)
        }
    }

    async function getRememberInfo() {
        const rememberInfo = getCookie(cookies.remember.name)

        if (!rememberInfo) {
            return
        }

        const { username, password } = rememberInfo

        dispatch(onSetRememberMe(true))
        dispatch(onChangeUsername(username))
        dispatch(onChangePassword(decrypt(password)))

    }

    const onToggleRememberMe = () => {
        dispatch(onChangeRememberMe())
    }

    const onToggleVisiblePassword = () => {
        dispatch(onToggleVisiblePasswordAction())
    }

    return (
        <AuthContext.Provider
            value={{
                user,
                isAuthenticated,
                rememberMe,
                isLoading,
                password,
                username,
                visiblePassword,
                onToggleRememberMe,
                onToggleVisiblePassword,
                signIn,
            }}
        >
            {children}
        </AuthContext.Provider>
    );
}

