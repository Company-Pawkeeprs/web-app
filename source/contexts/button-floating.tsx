import { usePathname } from 'next/navigation'
import { useEffect } from 'react'
import { create } from 'zustand'

interface ButtonFloatingContext {
    hasButtonFloating: boolean
    setHasButtonFloating: (hasButtonFloating: boolean) => void
}

export const useZustandBtnFloating = create<ButtonFloatingContext>((set) => ({
    hasButtonFloating: false,
    setHasButtonFloating: (hasButtonFloating) => set({ hasButtonFloating }),
}))

const list = ['/veterinary/dashboard/appointments', '/feedback']

export const useBtnFloating = () => {
    const pathname = usePathname()
    const { hasButtonFloating, setHasButtonFloating } = useZustandBtnFloating()

    useEffect(() => {
        const permission = list.some((item) => pathname.startsWith(item))
        if (permission) {
            setHasButtonFloating(false)
        } else {
            setHasButtonFloating(true)
        }
    }, [pathname])

    return {
        hasButtonFloating,
        setHasButtonFloating,
    }
}
