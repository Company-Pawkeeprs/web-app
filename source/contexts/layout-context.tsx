import ParticlesAuth from "~/Components/templates/particles-auth";

import { usePathname } from 'next/navigation';
import VerticalLayouts from "~/Layouts";

type LayoutProps = {
    children: React.ReactNode;
}

const routesAuth = ['/sign-in', '/sign-up', '/forgot-password', '/reset-password', '/logout']

const LayoutAuth = ({ children }: LayoutProps) => {
    return (
        <ParticlesAuth>
            {children}
        </ParticlesAuth>
    )
}

const LayoutDefault = ({ children }: LayoutProps) => {
    const pathname = usePathname()
    if (pathname === '/' || !pathname) {
        return (
            <>{children}</>
        )
    }

    if (routesAuth.includes(pathname as string)) {
        return <LayoutAuth>{children}</LayoutAuth>
    }

    return (
        <VerticalLayouts>
            {children as any}
        </VerticalLayouts>
    )
};

export default LayoutDefault;