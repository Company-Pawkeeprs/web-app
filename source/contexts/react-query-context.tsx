'use client'

import { QueryClient, onlineManager } from '@tanstack/react-query'
import { ReactQueryDevtools } from '@tanstack/react-query-devtools'

import { createSyncStoragePersister } from '@tanstack/query-sync-storage-persister'
import { PersistQueryClientProvider } from '@tanstack/react-query-persist-client'
import localforage from 'localforage'
import { useEffect } from 'react'

const queryClient = new QueryClient({
    defaultOptions: {
        queries: {
            networkMode: 'offlineFirst',
        },
    },
})

// Configuração do persister usando localForage
const localStoragePersister = createSyncStoragePersister({
    storage: localforage,
    retry: false,
})

type ReactQueryContextProps = {
    children: React.ReactNode
}

// Configuração do onlineManager para detectar status de rede
function setupOnlineManager() {
    if (typeof window !== 'undefined' && window.addEventListener) {
        const handleOnline = () => onlineManager.setOnline(true)
        const handleOffline = () => onlineManager.setOnline(false)

        window.addEventListener('online', handleOnline)
        window.addEventListener('offline', handleOffline)

        return () => {
            window.removeEventListener('online', handleOnline)
            window.removeEventListener('offline', handleOffline)
        }
    }
}

const ReactQueryContext = ({ children }: ReactQueryContextProps) => {
    useEffect(() => {
        setupOnlineManager()
    }, [])

    return (
        <PersistQueryClientProvider
            client={queryClient}
            persistOptions={{ persister: localStoragePersister }}
            onSuccess={() => {
                // Optional callback when hydration completes
                queryClient.resumePausedMutations()
            }}
        >
            {children}
            {process.env.NODE_ENV !== 'production' && (
                <ReactQueryDevtools buttonPosition="bottom-left" initialIsOpen />
            )}
        </PersistQueryClientProvider>
    )
}

export default ReactQueryContext
