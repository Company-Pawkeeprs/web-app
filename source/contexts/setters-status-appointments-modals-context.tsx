import { create } from 'zustand'
import CanceledScheduledModal from '~/Components/modals/canceled-scheduled-modal'
import ConfirmedScheduledModal from '~/Components/modals/confirmed-scheduled-modal'
import ReScheduledModal from '~/Components/modals/re-scheduled-modal'
import type { VeterinaryConsultation } from '~/types/appointment'

export enum ModalPlus {
    CanceledScheduled = 'CanceledScheduled',
    Rescheduled = 'Rescheduled',
    ConfirmedScheduled = 'ConfirmedScheduled',
}

interface PlusModalState {
    isOpen: {
        [key in ModalPlus]?: boolean
    }
    open: (key: ModalPlus) => void
    close: (key: ModalPlus) => void
    item: VeterinaryConsultation
    setItem: (item: VeterinaryConsultation) => void
    keys: typeof ModalPlus
}

export const usePlusModal = create<PlusModalState>((set) => ({
    isOpen: {},
    open: (key: ModalPlus) => set({ isOpen: { [key]: true } }),
    close: (key: ModalPlus) => set({ isOpen: { [key]: false } }),
    item: {} as VeterinaryConsultation,
    setItem: (item: VeterinaryConsultation) => set({ item }),
    keys: ModalPlus,
}))

const ContextSettersStatusAppointmentsModals = () => {
    const { isOpen, close, open } = usePlusModal()

    return (
        <>
            <CanceledScheduledModal
                closeModal={() => close(ModalPlus.CanceledScheduled)}
                isOpen={isOpen[ModalPlus.CanceledScheduled]}
                showModal={() => open(ModalPlus.CanceledScheduled)}
            >
                {() => <div className="hidden" />}
            </CanceledScheduledModal>

            <ReScheduledModal
                closeModal={() => close(ModalPlus.Rescheduled)}
                isOpen={isOpen[ModalPlus.Rescheduled]}
                showModal={() => open(ModalPlus.Rescheduled)}
            >
                {() => <div className="hidden" />}
            </ReScheduledModal>

            <ConfirmedScheduledModal
                closeModal={() => close(ModalPlus.ConfirmedScheduled)}
                isOpen={isOpen[ModalPlus.ConfirmedScheduled]}
                showModal={() => open(ModalPlus.ConfirmedScheduled)}
            >
                {() => <div className="hidden" />}
            </ConfirmedScheduledModal>
        </>
    )
}

export default ContextSettersStatusAppointmentsModals
