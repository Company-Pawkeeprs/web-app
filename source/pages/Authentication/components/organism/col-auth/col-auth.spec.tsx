import { render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';
import { useAuth } from '~/contexts/auth-context';
import ColAuth from './col-auth';

describe('ColAuth (Unit)', () => {
    it('should render the component', async () => {
        vi.mock('~/contexts/auth-context', () => ({
            useAuth: () => ({
                signIn: () => { },
                password: '',
                username: '',
                onToggleRememberMe: () => { },
                onToggleVisiblePassword: () => { },
                rememberMe: false,
                visiblePassword: false,
            }),
        }));

        render(<ColAuth />);

        const usernameInput = screen.getByLabelText('Nome do UsuÃ¡rio:');
        const passwordInput = screen.getByLabelText('Senha');
        const rememberMeInput = screen.getByLabelText('Lembrar-me');
        const submitButton = screen.getByText('Entrar');

        await userEvent.type(usernameInput, 'testuser');
        await userEvent.type(passwordInput, 'testpassword');
        await userEvent.click(rememberMeInput);
        await userEvent.click(submitButton);

        await waitFor(() => expect(useAuth().signIn).toHaveBeenCalled());
    });

    it('should toggle password visibility', () => {
        render(<ColAuth />);

        const passwordInput = screen.getByLabelText('Senha');
        const toggleButton = screen.getByTitle('Toggle password visibility');

        expect(passwordInput).toHaveAttribute('type', 'password');
        userEvent.click(toggleButton);
        expect(passwordInput).toHaveAttribute('type', 'text');
        userEvent.click(toggleButton);
        expect(passwordInput).toHaveAttribute('type', 'password');
    });

    it('should remember user when checkbox is checked', () => {
        render(<ColAuth />);
        const rememberCheckbox = screen.getByLabelText('Lembrar-me');

        expect(rememberCheckbox).not.toBeChecked();
        userEvent.click(rememberCheckbox);
        expect(rememberCheckbox).toBeChecked();
    });
});