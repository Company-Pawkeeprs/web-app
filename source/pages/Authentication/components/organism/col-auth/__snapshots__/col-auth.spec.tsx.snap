// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ColAuth Unit (Unit) > should render without errors 1`] = `
<body>
  <div>
    <div
      class="mb-3"
    >
      <label
        class="mb-0 list-group-item fs-12 form-label"
        data-testid="label-username"
        for="username"
      >
        Nome do Usuário:
        <span
          class="text-danger"
        >
          *
        </span>
      </label>
      <div
        class="position-relative mb-2 input-group"
      >
        <input
          class="form-control form-control"
          data-testid="username-input"
          id="username"
          name="username"
          placeholder="Digite seu nome de usuário"
          required=""
          type="text"
          value=""
        />
      </div>
      <label
        class="font-bold list-group-item text-danger fs-11 text-center"
        data-testid="err-username"
        style="display: none;"
      />
    </div>
    <div
      class="mb-3"
    >
      <div
        class="float-end"
      >
        <a
          class="text-muted"
          href="/forget-password"
        >
          Esqueceu a Senha?
        </a>
      </div>
      <label
        class="form-label form-label"
        for="password-input"
      >
        Senha
      </label>
      <div
        class="position-relative auth-pass-inputgroup mb-3"
      >
        <div
          class="position-relative mb-2 input-group"
        >
          <input
            class="form-control pe-5 password-input border-end-0 form-control"
            data-testid="password-input"
            id="password"
            name="password"
            placeholder="Digite sua senha"
            required=""
            type="password"
            value=""
          />
          <span
            class="bg-transparent border-start-0 input-group-text"
          >
            <i
              class="ri-eye-fill"
              data-testid="toggle-password"
            />
          </span>
        </div>
        <label
          class="font-bold list-group-item text-danger fs-11 text-center"
          data-testid="err-password"
          style="display: none;"
        />
      </div>
    </div>
    <div
      class="w-100 form-check"
    >
      <input
        class="form-check-input"
        id="rememberMe"
        name="rememberMe"
        type="checkbox"
      />
      <label
        class="form-check-label"
        for="rememberMe"
        title=""
      >
        <p
          class="mb-0 fs-12 text-muted fst-italic"
        >
          Lembrar-me
        </p>
      </label>
    </div>
    <div
      class="mt-4"
    >
      <button
        class="w-100 btn btn-primary"
        color="success"
        data-testid="submit-button"
        type="submit"
      >
        Entrar
      </button>
    </div>
  </div>
</body>
`;
